core/runtime/v2/shim_manager.go

```diff
@@ -163,37 +164,68 @@ func (m *ShimManager) ID() string {
 
 // Start launches a new shim instance
 func (m *ShimManager) Start(ctx context.Context, id string, bundle *Bundle, opts runtime.CreateOpts) (_ ShimInstance, retErr error) {
-	// This container belongs to sandbox which supposed to be already started via sandbox API.
+	shouldInvokeShimBinary := false
+
+	var params shimbinary.BootstrapParams
 	if opts.SandboxID != "" {
-		var params shimbinary.BootstrapParams
-		if opts.Address != "" {
-			// The address returned from sandbox controller should be in the form like ttrpc+unix://<uds-path>
-			// or grpc+vsock://<cid>:<port>, we should get the protocol from the url first.
-			protocol, address, ok := strings.Cut(opts.Address, "+")
-			if !ok {
-				return nil, fmt.Errorf("the scheme of sandbox address should be in " +
-					" the form of <protocol>+<unix|vsock|tcp>, i.e. ttrpc+unix or grpc+vsock")
-			}
-			params = shimbinary.BootstrapParams{
-				Version:  int(opts.Version),
-				Protocol: protocol,
-				Address:  address,
+		_, sbErr := m.sandboxStore.Get(ctx, opts.SandboxID)
+		if sbErr != nil {
+			if !errors.Is(sbErr, errdefs.ErrNotFound) {
+				return nil, sbErr
 			}
+
+			log.G(ctx).WithField("id", id).Warningf("sandbox (id=%s) not found, maybe created from v1.x", opts.SandboxID)
+			// NOTE: If sandbox container, like pause, is created by
+			// v1.6.x or v1.7.x, the shim may be not able to group
+			// multiple containers. We should invoke shim binary and
+			// establish new connection based on returned address.
+			shouldInvokeShimBinary = true
 		} else {
-			// For those sandbox we can not get endpoint,
-			// fallback to legacy implementation
-			process, err := m.Get(ctx, opts.SandboxID)
-			if err != nil {
-				return nil, fmt.Errorf("can't find sandbox %s", opts.SandboxID)
-			}
-			p, err := restoreBootstrapParams(process.Bundle())
-			if err != nil {
-				return nil, fmt.Errorf("failed to get bootstrap "+
-					"params of sandbox %s, legacy restore error %v", opts.SandboxID, err)
+			if opts.Address != "" {
+				// The address returned from sandbox controller should
+				// be in the form like ttrpc+unix://<uds-path> or grpc+vsock://<cid>:<port>,
+				// we should get the protocol from the url first.
+				protocol, address, ok := strings.Cut(opts.Address, "+")
+				if !ok {
+					return nil, fmt.Errorf("the scheme of sandbox address should be in " +
+						" the form of <protocol>+<unix|vsock|tcp>, i.e. ttrpc+unix or grpc+vsock")
+				}
+				params = shimbinary.BootstrapParams{
+					Version:  int(opts.Version),
+					Protocol: protocol,
+					Address:  address,
```

Comment:
I wonder if we should introduce a concept of "supported features" concept on the shim side. The bootstrap parameters could return a list of supported features as strings (e.g., `["sandboxes", "streaming_io", ...]`). This would allow containerd to explicitly check for feature support, rather than relying on assumptions.

Reply:
it sounds good to me. we can re-design this part. I think it's more easy to decide which client version we should use. but right now, version number is good enough to fix this compatible issue.